### ISSUE TITLE
CSRF issues with authentication within the Shopify embedded application

### ISSUE THREAD

**Issue by D0m-inic** [ericdude4/shopifex#75]
Hi, I've been having issues with CSRF and authentication while using this library.

My application has an added level of authentication, which requires logging in to a separate account within the embedded Shopify application. However, i've noticed that on trying to do a HTTP POST request, I get the error message:

```
invalid CSRF (Cross Site Request Forgery) token, please make sure that:

  * The session cookie is being sent and session is loaded
  * The request include a valid '_csrf_token' param or 'x-csrf-token' header
  ```
  
  However, I see that within the parameters, the _csrf_token is there:
  ```
  [info] POST /shopify/users/log_in
[debug] Processing with SoidWeb.UserSessionController.create/2
  Parameters: %{"_csrf_token" => "[redacted]", "token" => "[redacted]", "user" => %{"email" => "test@test.com", "password" => "[FILTERED]", "remember_me" => "false"}}
  ```
  
  My pipelines are set up in a relatively standard way:
  ```
    pipeline :browser do
    plug :accepts, ["html"]
    plug :fetch_session
    plug :fetch_live_flash
    plug :put_root_layout, html: {SoidWeb.Layouts, :root}
    plug :protect_from_forgery
    plug :put_secure_browser_headers
    plug :fetch_current_user
  end

  pipeline :shopify do
    plug :shopify_session
    plug :shopify_embedded
  end
  ```
  
  And the form in question which causes issues is the default phoenix phx.gen.auth controller authentication:
  ```
  <.simple_form :let={f} for={@conn.params["user"]} as={:user} action={"#{Soid.UrlManager.url(@conn,"/users/log_in")}"}>
    <.error :if={@error_message}><%= @error_message %></.error>

    <.input field={f[:email]} type="email" label="Email" required />
    <.input field={f[:password]} type="password" label="Password" required />

    <:actions :let={f}>
      <.input field={f[:remember_me]} type="checkbox" label="Keep me logged in" />
      <.link href={"#{Soid.UrlManager.url(@conn,"/users/reset_password")}"} class="text-sm font-semibold">
        Forgot your password?
      </.link>
    </:actions>
    <:actions>
      <.button phx-disable-with="Logging in..." class="w-full">
        Log in <span aria-hidden="true">→</span>
      </.button>
    </:actions>
  </.simple_form>
  ```
  
  Note that URLManager adds the JWT token to the params like so
  `"/shopify#{url}?token=#{Shopifex.Plug.session_token(conn)}"`
  (I'm not sure if there is a better way for this, please inform me if there is)
  
  I have tried adding the csrf token to the form using
  `<input type="hidden" name="_csrf_token" value={Plug.CSRFProtection.get_csrf_token()} />`
   however it gives the same issue.
  
  I believe this is an issue with conflicting CSRF tokens returned, however I cannot be sure. Any opinions are welcomed. Thanks

**Comment by D0m-inic** [comment]
Hi @ericdude4
Thanks for the quick response!

Unfortunately changing it to this doesn't seem to work still.
My application is ported as a Webapp and as a Shopify app, which allows me to compare the two and gives me some debugging information which I will provide here, as the Webapp application works fine without any CSRF issues.

Looking at how Plug.CSRFProtection works, it seems that the issue with the Shopify app stems from the **Plug.session** not having a csrf token.
The stem of the error comes from `get_session(conn, "_csrf_token")` returning nil.

On creating a POST request to /users/log_in:

For the Webapp:
the csrf_token= a full token
For the Shopify app:
the csrf_token= nil

Looking more closely at the `conn` struct for both at the GET request to /users/log_in, just before the POST request where the issue is, you can indeed see that the csrf token is in the webapp and not the shopify app within `Plug.Session`:

For the Webapp:
```
:plug_session => %{
      "_csrf_token" => "[token]",
      "user_return_to" => ...
    }
```

For the Shopify app:
```
:plug_session => %{
      "exp" => ...,
      "jwt" => "[redacted]"
    }
```

Looking closer, it appears that the issue stems from the cookies key within `conn`. Within this, the Webapp has the relevant key to sign cookies, as specified by default in endpoint.ex within the `@session_options`, whilst the Shopify app does not. I'm assuming that, as a result, cookies cannot be stored in the Shopify app, which causes the csrf token to not be stored in `Plug.Session`.

For the Webapp:
```
cookies: %{
    "_[appname]_key" => "[redacted]"
  }
  ```

For the shopify app:
```
cookies: %{"abuse_interstitial" => "[ngrok link]"}
```

This leads me to believe that perhaps the issue stems from ngrok not allowing cookies or the Set-cookie header not being correctly set due to issues with ngrok.
The fix for this, however, is still in question. Please let me know if you have any ideas or comments.

Thanks

**Comment by ericdude4** [comment]
Hi @D0m-inic,

Yes, cookies do not work in Shopify's embedded iFrame environment unless they are secure. I have my `endpoint.ex` file configured with the following:

```
  @session_options [
    store: :cookie,
    key: "_foo_session",
    signing_salt: "TeStIn9",
    secure: true,
    extra: "SameSite=None"
  ]

  ...

  plug Plug.Session, @session_options
```

Perhaps this difference will help?

Eric

**Comment by D0m-inic** [comment]
Yes this works now, thanks. Are there any safety concerns with this approach?

**Comment by ericdude4** [comment]
As far as I know, it's completely safe. It's just a more secure way of storing cookies - which is required by Shopify in order to use cookies at all.
